File: board/native/kvs_master/src/h264.rs
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AccessUnit {
    pub bytes: Vec<u8>,
    pub is_keyframe: bool,
}

#[derive(Debug, Default)]
pub struct AnnexBAccessUnitParser {
    pending: Vec<u8>,
    current: Vec<u8>,
    current_has_vcl: bool,
    current_is_keyframe: bool,
}

impl AnnexBAccessUnitParser {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn push(&mut self, chunk: &[u8]) -> Vec<AccessUnit> {
        self.pending.extend_from_slice(chunk);
        let mut output = Vec::new();
        let nals = extract_complete_nals(&mut self.pending);
        for nal in nals {
            self.push_nal(&nal, &mut output);
        }
        output
    }

    pub fn finish(&mut self) -> Vec<AccessUnit> {
        let mut output = Vec::new();
        if let Some(trailing) = take_trailing_nal(&mut self.pending) {
            self.push_nal(&trailing, &mut output);
        }
        self.flush_current(&mut output);
        output
    }

    fn push_nal(&mut self, nal: &[u8], output: &mut Vec<AccessUnit>) {
        let Some(payload) = nal_payload(nal) else {
            return;
        };
        let nal_type = payload[0] & 0x1f;
        let is_vcl = (1..=5).contains(&nal_type);

        if nal_type == 9 {
            self.flush_current(output);
            return;
        }

        if is_vcl && self.current_has_vcl && first_mb_in_slice_is_zero(payload) {
            self.flush_current(output);
        } else if !is_vcl && self.current_has_vcl && matches!(nal_type, 6 | 7 | 8) {
            self.flush_current(output);
        }

        self.current.extend_from_slice(nal);
        if is_vcl {
            self.current_has_vcl = true;
            if nal_type == 5 {
                self.current_is_keyframe = true;
            }
        }
    }

    fn flush_current(&mut self, output: &mut Vec<AccessUnit>) {
        if self.current_has_vcl && !self.current.is_empty() {
            output.push(AccessUnit {
                bytes: std::mem::take(&mut self.current),
                is_keyframe: self.current_is_keyframe,
            });
        } else {
            self.current.clear();
        }

        self.current_has_vcl = false;
        self.current_is_keyframe = false;
    }
}

fn extract_complete_nals(pending: &mut Vec<u8>) -> Vec<Vec<u8>> {
    if pending.is_empty() {
        return Vec::new();
    }

    let Some((first_start, _)) = find_start_code(pending, 0) else {
        if pending.len() > 4 {
            let keep_from = pending.len() - 4;
            pending.drain(..keep_from);
        }
        return Vec::new();
    };

    if first_start > 0 {
        pending.drain(..first_start);
    }

    let mut starts = Vec::new();
    let mut index = 0;
    while let Some((start, prefix_len)) = find_start_code(pending, index) {
        starts.push((start, prefix_len));
        index = start + prefix_len;
    }

    if starts.len() < 2 {
        return Vec::new();
    }

    let mut nals = Vec::with_capacity(starts.len() - 1);
    for window in starts.windows(2) {
        let start = window[0].0;
        let end = window[1].0;
        nals.push(pending[start..end].to_vec());
    }

    let tail_start = starts
        .last()
        .map(|(start, _)| *start)
        .expect("tail_start requires at least one start code");
    let remainder = pending[tail_start..].to_vec();
    *pending = remainder;
    nals
}

fn take_trailing_nal(pending: &mut Vec<u8>) -> Option<Vec<u8>> {
    let (start, prefix_len) = find_start_code(pending, 0)?;
    if start != 0 || pending.len() <= prefix_len {
        pending.clear();
        return None;
    }
    Some(std::mem::take(pending))
}

fn nal_payload(nal: &[u8]) -> Option<&[u8]> {
    let (_, prefix_len) = find_start_code(nal, 0)?;
    nal.get(prefix_len..)
}

fn find_start_code(data: &[u8], from: usize) -> Option<(usize, usize)> {
    if data.len() < 4 || from >= data.len().saturating_sub(3) {
        return None;
    }

    let mut index = from;
    while index + 3 < data.len() {
        if data[index] == 0 && data[index + 1] == 0 {
            if data[index + 2] == 1 {
                return Some((index, 3));
            }
            if index + 3 < data.len() && data[index + 2] == 0 && data[index + 3] == 1 {
                return Some((index, 4));
            }
        }
        index += 1;
    }
    None
}

fn first_mb_in_slice_is_zero(payload: &[u8]) -> bool {
    if payload.len() < 2 {
        return false;
    }

    let rbsp = remove_emulation_prevention_bytes(&payload[1..]);
    let mut reader = BitReader::new(&rbsp);
    matches!(reader.read_ue(), Some(0))
}

fn remove_emulation_prevention_bytes(payload: &[u8]) -> Vec<u8> {
    let mut rbsp = Vec::with_capacity(payload.len());
    let mut zero_run = 0_u8;

    for byte in payload {
        if zero_run >= 2 && *byte == 0x03 {
            zero_run = 0;
            continue;
        }
        rbsp.push(*byte);
        if *byte == 0 {
            zero_run = zero_run.saturating_add(1);
        } else {
            zero_run = 0;
        }
    }

    rbsp
}

struct BitReader<'a> {
    bytes: &'a [u8],
    bit_offset: usize,
}

impl<'a> BitReader<'a> {
    fn new(bytes: &'a [u8]) -> Self {
        Self {
            bytes,
            bit_offset: 0,
        }
    }

    fn read_bit(&mut self) -> Option<u8> {
        let byte = *self.bytes.get(self.bit_offset / 8)?;
        let shift = 7 - (self.bit_offset % 8);
        self.bit_offset += 1;
        Some((byte >> shift) & 0x01)
    }

    fn read_bits(&mut self, count: usize) -> Option<u32> {
        let mut value = 0_u32;
        for _ in 0..count {
            value = (value << 1) | u32::from(self.read_bit()?);
        }
        Some(value)
    }

    fn read_ue(&mut self) -> Option<u32> {
        let mut leading_zeros = 0_usize;
        while self.read_bit()? == 0 {
            leading_zeros += 1;
        }
        if leading_zeros == 0 {
            return Some(0);
        }
        let suffix = self.read_bits(leading_zeros)?;
        Some(((1_u32 << leading_zeros) - 1) + suffix)
    }
}

#[cfg(test)]
mod tests {
    use super::{AccessUnit, AnnexBAccessUnitParser};

    fn nal(bytes: &[u8]) -> Vec<u8> {
        let mut data = vec![0x00, 0x00, 0x00, 0x01];
        data.extend_from_slice(bytes);
        data
    }

    #[test]
    fn assembles_access_units_from_chunked_annex_b_stream() {
        let mut parser = AnnexBAccessUnitParser::new();
        let mut stream = Vec::new();
        stream.extend(nal(&[0x67, 0xaa]));
        stream.extend(nal(&[0x68, 0xbb]));
        stream.extend(nal(&[0x65, 0x80]));
        stream.extend(nal(&[0x41, 0x80]));

        let mut access_units = Vec::new();
        access_units.extend(parser.push(&stream[..11]));
        access_units.extend(parser.push(&stream[11..]));
        access_units.extend(parser.finish());

        assert_eq!(
            access_units,
            vec![
                AccessUnit {
                    bytes: [nal(&[0x67, 0xaa]), nal(&[0x68, 0xbb]), nal(&[0x65, 0x80])].concat(),
                    is_keyframe: true,
                },
                AccessUnit {
                    bytes: nal(&[0x41, 0x80]),
                    is_keyframe: false,
                },
            ]
        );
    }

    #[test]
    fn treats_aud_as_boundary() {
        let mut parser = AnnexBAccessUnitParser::new();
        let mut stream = Vec::new();
        stream.extend(nal(&[0x09, 0xf0]));
        stream.extend(nal(&[0x65, 0x80]));
        stream.extend(nal(&[0x09, 0xf0]));
        stream.extend(nal(&[0x41, 0x80]));

        let mut access_units = parser.push(&stream);
        access_units.extend(parser.finish());

        assert_eq!(access_units.len(), 2);
        assert!(access_units[0].is_keyframe);
        assert!(!access_units[1].is_keyframe);
    }
}

File: board/kvs_master/src/h264.cpp
#include "txing_board_kvs_master/h264.hpp"

#include <optional>

namespace txing::board::kvs_master {
namespace {

using StartCode = std::pair<std::size_t, std::size_t>;

std::optional<StartCode> FindStartCode(const std::vector<std::uint8_t>& data, std::size_t from) {
    if (data.size() < 4 || from >= data.size() - 3) {
        return std::nullopt;
    }

    for (std::size_t index = from; index + 3 < data.size(); ++index) {
        if (data[index] == 0x00 && data[index + 1] == 0x00) {
            if (data[index + 2] == 0x01) {
                return StartCode(index, 3);
            }
            if (index + 3 < data.size() && data[index + 2] == 0x00 && data[index + 3] == 0x01) {
                return StartCode(index, 4);
            }
        }
    }
    return std::nullopt;
}

std::optional<std::vector<std::uint8_t>> TakeTrailingNal(std::vector<std::uint8_t>& pending) {
    const auto start = FindStartCode(pending, 0);
    if (!start || start->first != 0 || pending.size() <= start->second) {
        pending.clear();
        return std::nullopt;
    }

    auto trailing = std::move(pending);
    pending.clear();
    return trailing;
}

std::vector<std::vector<std::uint8_t>> ExtractCompleteNals(std::vector<std::uint8_t>& pending) {
    if (pending.empty()) {
        return {};
    }

    const auto first_start = FindStartCode(pending, 0);
    if (!first_start) {
        if (pending.size() > 4) {
            pending.erase(pending.begin(), pending.end() - 4);
        }
        return {};
    }

    if (first_start->first > 0) {
        pending.erase(pending.begin(), pending.begin() + static_cast<std::ptrdiff_t>(first_start->first));
    }

    std::vector<StartCode> starts;
    std::size_t index = 0;
    while (const auto start = FindStartCode(pending, index)) {
        starts.push_back(*start);
        index = start->first + start->second;
    }

    if (starts.size() < 2) {
        return {};
    }

    std::vector<std::vector<std::uint8_t>> nals;
    nals.reserve(starts.size() - 1);
    for (std::size_t i = 0; i + 1 < starts.size(); ++i) {
        const auto start = starts[i].first;
        const auto end = starts[i + 1].first;
        nals.emplace_back(pending.begin() + static_cast<std::ptrdiff_t>(start), pending.begin() + static_cast<std::ptrdiff_t>(end));
    }

    const auto tail_start = starts.back().first;
    pending.erase(pending.begin(), pending.begin() + static_cast<std::ptrdiff_t>(tail_start));
    return nals;
}

const std::uint8_t* NalPayload(const std::vector<std::uint8_t>& nal, std::size_t& payload_size) {
    const auto start = FindStartCode(nal, 0);
    if (!start || nal.size() <= start->second) {
        payload_size = 0;
        return nullptr;
    }
    payload_size = nal.size() - start->second;
    return nal.data() + static_cast<std::ptrdiff_t>(start->second);
}

std::vector<std::uint8_t> RemoveEmulationPreventionBytes(const std::uint8_t* payload, std::size_t size) {
    std::vector<std::uint8_t> rbsp;
    rbsp.reserve(size);
    std::uint8_t zero_run = 0;
    for (std::size_t index = 0; index < size; ++index) {
        const auto byte = payload[index];
        if (zero_run >= 2 && byte == 0x03) {
            zero_run = 0;
            continue;
        }
        rbsp.push_back(byte);
        if (byte == 0x00) {
            zero_run = static_cast<std::uint8_t>(zero_run + 1);
        } else {
            zero_run = 0;
        }
    }
    return rbsp;
}

class BitReader {
  public:
    explicit BitReader(const std::vector<std::uint8_t>& bytes) : bytes_(bytes) {}

    std::optional<std::uint8_t> ReadBit() {
        if (bit_offset_ / 8 >= bytes_.size()) {
            return std::nullopt;
        }

        const auto byte = bytes_[bit_offset_ / 8];
        const auto shift = 7U - static_cast<unsigned>(bit_offset_ % 8);
        ++bit_offset_;
        return static_cast<std::uint8_t>((byte >> shift) & 0x01U);
    }

    std::optional<std::uint32_t> ReadBits(std::size_t count) {
        std::uint32_t value = 0;
        for (std::size_t index = 0; index < count; ++index) {
            const auto bit = ReadBit();
            if (!bit) {
                return std::nullopt;
            }
            value = (value << 1U) | static_cast<std::uint32_t>(*bit);
        }
        return value;
    }

    std::optional<std::uint32_t> ReadUnsignedExpGolomb() {
        std::size_t leading_zeros = 0;
        while (true) {
            const auto bit = ReadBit();
            if (!bit) {
                return std::nullopt;
            }
            if (*bit == 1U) {
                break;
            }
            ++leading_zeros;
        }

        if (leading_zeros == 0) {
            return 0;
        }

        const auto suffix = ReadBits(leading_zeros);
        if (!suffix) {
            return std::nullopt;
        }
        return ((1U << leading_zeros) - 1U) + *suffix;
    }

  private:
    const std::vector<std::uint8_t>& bytes_;
    std::size_t bit_offset_ = 0;
};

bool FirstMbInSliceIsZero(const std::uint8_t* payload, std::size_t size) {
    if (size < 2) {
        return false;
    }
    const auto rbsp = RemoveEmulationPreventionBytes(payload + 1, size - 1);
    BitReader reader(rbsp);
    const auto first_mb = reader.ReadUnsignedExpGolomb();
    return first_mb && *first_mb == 0;
}

}  // namespace

bool AccessUnit::operator==(const AccessUnit& other) const {
    return is_keyframe == other.is_keyframe && bytes == other.bytes;
}

std::vector<AccessUnit> AnnexBAccessUnitParser::Push(const std::uint8_t* chunk, std::size_t size) {
    pending_.insert(pending_.end(), chunk, chunk + static_cast<std::ptrdiff_t>(size));

    std::vector<AccessUnit> output;
    const auto nals = ExtractCompleteNals(pending_);
    for (const auto& nal : nals) {
        PushNal(nal, output);
    }
    return output;
}

std::vector<AccessUnit> AnnexBAccessUnitParser::Finish() {
    std::vector<AccessUnit> output;
    if (const auto trailing = TakeTrailingNal(pending_)) {
        PushNal(*trailing, output);
    }
    FlushCurrent(output);
    return output;
}

void AnnexBAccessUnitParser::PushNal(const std::vector<std::uint8_t>& nal, std::vector<AccessUnit>& output) {
    std::size_t payload_size = 0;
    const auto* payload = NalPayload(nal, payload_size);
    if (payload == nullptr || payload_size == 0) {
        return;
    }

    const auto nal_type = static_cast<std::uint8_t>(payload[0] & 0x1fU);
    const bool is_vcl = nal_type >= 1 && nal_type <= 5;

    if (nal_type == 9) {
        FlushCurrent(output);
        return;
    }

    if (is_vcl && current_has_vcl_ && FirstMbInSliceIsZero(payload, payload_size)) {
        FlushCurrent(output);
    } else if (!is_vcl && current_has_vcl_ && (nal_type == 6 || nal_type == 7 || nal_type == 8)) {
        FlushCurrent(output);
    }

    current_.insert(current_.end(), nal.begin(), nal.end());
    if (is_vcl) {
        current_has_vcl_ = true;
        if (nal_type == 5) {
            current_is_keyframe_ = true;
        }
    }
}

void AnnexBAccessUnitParser::FlushCurrent(std::vector<AccessUnit>& output) {
    if (current_has_vcl_ && !current_.empty()) {
        AccessUnit access_unit;
        access_unit.bytes = std::move(current_);
        access_unit.is_keyframe = current_is_keyframe_;
        output.push_back(std::move(access_unit));
    } else {
        current_.clear();
    }

    current_has_vcl_ = false;
    current_is_keyframe_ = false;
}

}  // namespace txing::board::kvs_master
